mmv1/third_party/terraform/services/bigquery/iam_bigquery_member_dataset.go

```diff
@@ -0,0 +1,322 @@
+	"dataset_id": {
+		Type:     schema.TypeString,
+		Required: true,
+		ForceNew: true,
+	},
+	"project": {
+		Type:     schema.TypeString,
+		Optional: true,
+		Computed: true,
+		ForceNew: true,
+	},
+}
+
+var bigqueryIamMemberAccessPrimitiveToRoleMap = map[string]string{
+	"OWNER":  "roles/bigquery.dataOwner",
+	"WRITER": "roles/bigquery.dataEditor",
+	"READER": "roles/bigquery.dataViewer",
+}
+
+type BigqueryDatasetIamMemberUpdater struct {
+	project   string
+	datasetId string
+	d         tpgresource.TerraformResourceData
+	Config    *transport_tpg.Config
+}
+
+func NewBigqueryDatasetIamMemberUpdater(d tpgresource.TerraformResourceData, config *transport_tpg.Config) (tpgiamresource.ResourceIamUpdater, error) {
+	project, err := tpgresource.GetProject(d, config)
+	if err != nil {
+		return nil, err
+	}
+
+	if err := d.Set("project", project); err != nil {
+		return nil, fmt.Errorf("Error setting project: %s", err)
+	}
+
+	return &BigqueryDatasetIamMemberUpdater{
+		project:   project,
+		datasetId: d.Get("dataset_id").(string),
+		d:         d,
+		Config:    config,
+	}, nil
+}
+
+func (u *BigqueryDatasetIamMemberUpdater) GetResourceIamPolicy() (*cloudresourcemanager.Policy, error) {
+	url := fmt.Sprintf("%s%s", u.Config.BigQueryBasePath, u.GetResourceId())
+
+	userAgent, err := tpgresource.GenerateUserAgentString(u.d, u.Config.UserAgent)
+	if err != nil {
+		return nil, err
+	}
+
+	res, err := transport_tpg.SendRequest(transport_tpg.SendRequestOptions{
+		Config:    u.Config,
+		Method:    "GET",
+		Project:   u.project,
+		RawURL:    url,
+		UserAgent: userAgent,
+	})
+	if err != nil {
+		return nil, errwrap.Wrapf(fmt.Sprintf("Error retrieving IAM policy for %s: {{err}}", u.DescribeResource()), err)
+	}
+
+	policy, err := accessToPolicyForIamMember(res["access"])
+	if err != nil {
+		return nil, err
+	}
+	return policy, nil
+}
+
+func GetCurrentResourceAccess(u *BigqueryDatasetIamMemberUpdater) ([]interface{}, error) {
+	url := fmt.Sprintf("%s%s", u.Config.BigQueryBasePath, u.GetResourceId())
+
+	userAgent, err := tpgresource.GenerateUserAgentString(u.d, u.Config.UserAgent)
+	if err != nil {
+		return nil, err
+	}
+
+	res, err := transport_tpg.SendRequest(transport_tpg.SendRequestOptions{
+		Config:    u.Config,
+		Method:    "GET",
+		Project:   u.project,
+		RawURL:    url,
+		UserAgent: userAgent,
+	})
+	if err != nil {
+		return nil, errwrap.Wrapf(fmt.Sprintf("Error retrieving IAM policy for %s: {{err}}", u.DescribeResource()), err)
+	}
+
+	var access []interface{}
+	if accessVal, ok := res["access"].([]interface{}); ok {
+		access = accessVal
+	} else if res["access"] == nil {
+		access = []interface{}{} // Return an empty slice if the key is missing
+	} else {
+		return nil, fmt.Errorf("value under 'access' is not a slice of interface{}")
+	}
+
+	return access, nil
+}
+
+func mergeAccess(newAccess []map[string]interface{}, currAccess []interface{}) []map[string]interface{} {
```

Comment:
This is a pretty basic merging logic, which doesn't check for duplicates.  The BQ API does seem to handle this. Do you think this can cause any issues in the terraform layer? @c2thorn .
E.x.
REQUEST:
```
 "access": [
  {
   "role": "roles/bigquery.dataEditor",
   "specialGroup": "projectWriters"
  },
  {
   "role": "roles/bigquery.dataOwner",
   "specialGroup": "projectOwners"
  },
  {
   "role": "roles/bigquery.dataOwner",
   "userByEmail": "<redacted>"
  },
  {
   "role": "roles/bigquery.dataViewer",
   "specialGroup": "projectReaders"
  },
  {
   "role": "roles/viewer",
   "userByEmail": "tf-test-apr40cg93z@<redacted>.iam.gserviceaccount.com"
  },
  {
   "role": "WRITER",
   "specialGroup": "projectWriters"
  },
  {
   "role": "OWNER",
   "specialGroup": "projectOwners"
  },
  {
   "role": "OWNER",
   "userByEmail": "<redacted>"
  },
  {
   "role": "READER",
   "specialGroup": "projectReaders"
  },
  {
   "dataset": {
    "dataset": {
     "datasetId": "tf_test_k36vxbgshy",
     "projectId": "<redacted>"
    },
    "targetTypes": [
     "VIEWS"
    ]
   }
  }
 ]
}
```

RESPONSE:

```
 "access": [
    {
      "role": "WRITER",
      "specialGroup": "projectWriters"
    },
    {
      "role": "OWNER",
      "specialGroup": "projectOwners"
    },
    {
      "role": "OWNER",
      "userByEmail": "<redacted>"
    },
    {
      "role": "READER",
      "specialGroup": "projectReaders"
    },
    {
      "role": "roles/viewer",
      "userByEmail": "tf-test-apr40cg93z@<redacted>.iam.gserviceaccount.com"
    },
    {
      "dataset": {
        "dataset": {
          "datasetId": "tf_test_k36vxbgshy",
          "projectId": "<redacted>"
        },
        "targetTypes": [
"VIEWS"
        ]
      }
    }
  ],
```

Reply:
Is your example meant to show duplicated entries? I can't seem to find exact duplicate entries